File: 19th/코딩테스트 기초반/week4/1935번_후위 표기식2.cpp
#include <iostream>
#include <stack>
#include <string>
#include <vector>

using namespace std;

int main() {
    int N;
    string str;
    stack<double> S;
    
    cin >> N;
    cin >> str; // 후위 표기식
    
    vector<double> num(N); // 크기 N으로 초기화
    for (int i = 0; i < N; i++) {
        cin >> num[i];
    }
       
    for (int i = 0; i < str.length(); i++) {
        if (str[i] >= 'A' && str[i] <= 'Z') { // 피연산자일 경우 push  
            S.push(num[str[i] - 'A']);
        }
        else { // 연산자일 경우 stack의 상위 두 원소 pop
            double ans, tmp1, tmp2;
            tmp1 = S.top();
            S.pop();
            tmp2 = S.top();
            S.pop();

            // 먼저 나온 피연산자가 앞으로 오도록 함
            if (str[i] == '+') ans = tmp2 + tmp1; 
            if (str[i] == '-') ans = tmp2 - tmp1;
            if (str[i] == '*') ans = tmp2 * tmp1;
            if (str[i] == '/') ans = tmp2 / tmp1;

            S.push(ans);
        }     
    }
    
    // 소수점 2자리까지 표시하기 위함
    cout << fixed;
    cout.precision(2);
    
    cout << S.top();
}

File: 19th/코딩테스트 기초반/week4/1935번_후위 표기식2.py
n = int(input())
str = list(input())
s = []  # stack을 사용하기 위해 리스트 선언

alphabet = [0] * 26  # 피연산자를 저장할 알파벳 배열
for i in range(n):
    alphabet[i] = int(input())

for i in str:
    if i >= 'A' and i <= 'Z':  # i가 알파벳 일 때 
        s.append(alphabet[ord(i) - 65])  # stack에 i에 해당하는 수 push (알파벳 순서를 알아내기 위해 ord(i) - 65)
    else:
        tmp1 = s.pop()
        tmp2 = s.pop()
        if i == '+':
            ans = tmp2 + tmp1
        elif i == '-':
            ans = tmp2 - tmp1
        elif i == '*':
            ans = tmp2 * tmp1
        else:
            ans = tmp2 / tmp1
        s.append(ans)

print(f"{s.pop():.2f}")  # 소수점 2자리까지 출력 